File: API/0483_Five_Ema_NoTouch_Breakout/CS/FiveEmaNoTouchBreakoutStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// 5 EMA No-Touch Breakout Strategy - waits for a candle completely above or below the EMA and enters on breakout.
/// </summary>
public class FiveEmaNoTouchBreakoutStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _emaPeriod;
	private readonly StrategyParam<decimal> _rewardRisk;
	private readonly StrategyParam<int> _cooldownBars;

	private ExponentialMovingAverage _ema;

	private decimal? _pendingLongHigh;
	private decimal? _pendingLongLow;
	private decimal? _pendingShortLow;
	private decimal? _pendingShortHigh;
	private bool _longReady;
	private bool _shortReady;

	private decimal? _longStop;
	private decimal? _longTarget;
	private decimal? _shortStop;
	private decimal? _shortTarget;
	private int _cooldownRemaining;

	/// <summary>
	/// Candle type for strategy calculation.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// EMA period.
	/// </summary>
	public int EmaPeriod
	{
		get => _emaPeriod.Value;
		set => _emaPeriod.Value = value;
	}

	/// <summary>
	/// Reward to risk ratio.
	/// </summary>
	public decimal RewardRisk
	{
		get => _rewardRisk.Value;
		set => _rewardRisk.Value = value;
	}

	public int CooldownBars
	{
		get => _cooldownBars.Value;
		set => _cooldownBars.Value = value;
	}

	/// <summary>
	/// Constructor.
	/// </summary>
	public FiveEmaNoTouchBreakoutStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(30).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles to use", "General");

		_emaPeriod = Param(nameof(EmaPeriod), 5)
			.SetGreaterThanZero()
			.SetDisplay("EMA Period", "Length of EMA", "EMA")
			
			.SetOptimize(3, 10, 1);

		_rewardRisk = Param(nameof(RewardRisk), 3.0m)
			.SetGreaterThanZero()
			.SetDisplay("Reward : Risk", "Reward to risk ratio", "Risk Management")
			
			.SetOptimize(1.0m, 5.0m, 0.5m);

		_cooldownBars = Param(nameof(CooldownBars), 10)
			.SetDisplay("Cooldown Bars", "Bars between trades", "Risk");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_pendingLongHigh = null;
		_pendingLongLow = null;
		_pendingShortLow = null;
		_pendingShortHigh = null;
		_longReady = false;
		_shortReady = false;
		_longStop = null;
		_longTarget = null;
		_shortStop = null;
		_shortTarget = null;
		_cooldownRemaining = 0;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		_ema = new ExponentialMovingAverage { Length = EmaPeriod };

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(_ema, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, _ema);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal emaValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading() || !_ema.IsFormed)
			return;

		var high = candle.HighPrice;
		var low = candle.LowPrice;
		var close = candle.ClosePrice;

		if (high < emaValue)
		{
			_pendingLongHigh = high;
			_pendingLongLow = low;
			_longReady = true;
			_shortReady = false;
		}
		else if (low > emaValue)
		{
			_pendingShortLow = low;
			_pendingShortHigh = high;
			_shortReady = true;
			_longReady = false;
		}

		// Check stop/target exits (always, regardless of cooldown)
		if (Position > 0 && _longStop is decimal ls && _longTarget is decimal lt)
		{
			if (low <= ls || high >= lt)
			{
				SellMarket(Math.Abs(Position));
				_longStop = null;
				_longTarget = null;
				_cooldownRemaining = CooldownBars;
				return;
			}
		}
		else if (Position < 0 && _shortStop is decimal ss && _shortTarget is decimal st)
		{
			if (high >= ss || low <= st)
			{
				BuyMarket(Math.Abs(Position));
				_shortStop = null;
				_shortTarget = null;
				_cooldownRemaining = CooldownBars;
				return;
			}
		}

		if (_cooldownRemaining > 0)
		{
			_cooldownRemaining--;
			return;
		}

		if (_longReady && _pendingLongHigh is decimal longHigh && high > longHigh)
		{
			if (_pendingLongLow is decimal longLow)
			{
				_longStop = longLow;
				_longTarget = close + (close - longLow) * RewardRisk;
				BuyMarket(Volume + Math.Abs(Position));
				_longReady = false;
				_cooldownRemaining = CooldownBars;
			}
		}
		else if (_shortReady && _pendingShortLow is decimal shortLow && low < shortLow)
		{
			if (_pendingShortHigh is decimal shortHigh)
			{
				_shortStop = shortHigh;
				_shortTarget = close - (shortHigh - close) * RewardRisk;
				SellMarket(Volume + Math.Abs(Position));
				_shortReady = false;
				_cooldownRemaining = CooldownBars;
			}
		}
	}
}

File: API/0483_Five_Ema_NoTouch_Breakout/PY/five_ema_no_touch_breakout_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan, Math
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy


class five_ema_no_touch_breakout_strategy(Strategy):
    """5 EMA No-Touch Breakout Strategy."""

    def __init__(self):
        super(five_ema_no_touch_breakout_strategy, self).__init__()

        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(30))) \
            .SetDisplay("Candle Type", "Type of candles to use", "General")
        self._ema_period = self.Param("EmaPeriod", 5) \
            .SetDisplay("EMA Period", "Length of EMA", "EMA")
        self._reward_risk = self.Param("RewardRisk", 3.0) \
            .SetDisplay("Reward : Risk", "Reward to risk ratio", "Risk Management")
        self._cooldown_bars = self.Param("CooldownBars", 10) \
            .SetDisplay("Cooldown Bars", "Bars between trades", "Risk")

        self._ema = None
        self._pending_long_high = None
        self._pending_long_low = None
        self._pending_short_low = None
        self._pending_short_high = None
        self._long_ready = False
        self._short_ready = False
        self._long_stop = None
        self._long_target = None
        self._short_stop = None
        self._short_target = None
        self._cooldown_remaining = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(five_ema_no_touch_breakout_strategy, self).OnReseted()
        self._ema = None
        self._pending_long_high = None
        self._pending_long_low = None
        self._pending_short_low = None
        self._pending_short_high = None
        self._long_ready = False
        self._short_ready = False
        self._long_stop = None
        self._long_target = None
        self._short_stop = None
        self._short_target = None
        self._cooldown_remaining = 0

    def OnStarted(self, time):
        super(five_ema_no_touch_breakout_strategy, self).OnStarted(time)

        self._ema = ExponentialMovingAverage()
        self._ema.Length = int(self._ema_period.Value)

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self._ema, self._on_process).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, self._ema)
            self.DrawOwnTrades(area)

    def _on_process(self, candle, ema_value):
        if candle.State != CandleStates.Finished:
            return

        if not self.IsFormedAndOnlineAndAllowTrading() or not self._ema.IsFormed:
            return

        high = float(candle.HighPrice)
        low = float(candle.LowPrice)
        close = float(candle.ClosePrice)
        ema_val = float(ema_value)
        cooldown = int(self._cooldown_bars.Value)
        rr = float(self._reward_risk.Value)

        if high < ema_val:
            self._pending_long_high = high
            self._pending_long_low = low
            self._long_ready = True
            self._short_ready = False
        elif low > ema_val:
            self._pending_short_low = low
            self._pending_short_high = high
            self._short_ready = True
            self._long_ready = False

        # Check stop/target exits
        if self.Position > 0 and self._long_stop is not None and self._long_target is not None:
            if low <= self._long_stop or high >= self._long_target:
                self.SellMarket(Math.Abs(self.Position))
                self._long_stop = None
                self._long_target = None
                self._cooldown_remaining = cooldown
                return
        elif self.Position < 0 and self._short_stop is not None and self._short_target is not None:
            if high >= self._short_stop or low <= self._short_target:
                self.BuyMarket(Math.Abs(self.Position))
                self._short_stop = None
                self._short_target = None
                self._cooldown_remaining = cooldown
                return

        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1
            return

        if self._long_ready and self._pending_long_high is not None and high > self._pending_long_high:
            if self._pending_long_low is not None:
                self._long_stop = self._pending_long_low
                self._long_target = close + (close - self._pending_long_low) * rr
                self.BuyMarket(self.Volume + Math.Abs(self.Position))
                self._long_ready = False
                self._cooldown_remaining = cooldown
        elif self._short_ready and self._pending_short_low is not None and low < self._pending_short_low:
            if self._pending_short_high is not None:
                self._short_stop = self._pending_short_high
                self._short_target = close - (self._pending_short_high - close) * rr
                self.SellMarket(self.Volume + Math.Abs(self.Position))
                self._short_ready = False
                self._cooldown_remaining = cooldown

    def CreateClone(self):
        return five_ema_no_touch_breakout_strategy()
